cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java

```diff
@@ -154,6 +160,120 @@ public ConnectorDag(Dag dag, Set<String> reduceNodes, Set<String> isolationNodes
       }
     }
 
+    /*
+        As an optimization, check if any connectors can be merged together.
+        It is generally more performant to merge connectors if possible, as it will result in less I/O.
+        To do this, we get the parents for all connectors, stopping at any source. Other connectors count as sources.
+        Connectors can be merged if their branch inputs are exactly the same. For example, consider the following dag:
+
+          s1 -----|  |--> t1 --> r1.connector --> r1 --> sink1
+                  |--|
+               |--|  |--> t2 --> r2.connector --> r2 --> sink2
+          s2 --|
+               |--> t3
+
+        The branch for r1.connector is:
+
+          t1 --> r1.connector --> r1 --> sink1
+
+        with branch inputs [source1, source2]. The branch for r2.connector is:
+
+          t2 --> r2.connector --> r2 --> sink2
+
+        with branch inputs [source1, source2]. Since the branches have the same inputs, they can be merged to:
+
+          s1 -----|                         |--> t1 --> r1 --> sink1
+                  |-- s1.s2.out.connector --|
+               |--|                         |--> t2  --> r2 --> sink2
+          s2 --|
+               |--> t3
+
+        which will eventually be split to:
+
+          s1 -----|                         s1.s2.out.connector --> t1 --> r1 --> sink1
+                  |-- s1.s2.out.connector
+               |--|                         s1.s2.out.connector --> t2 --> r2 --> sink2
+          s2 --|
+               |--> t3
+     */
+    Map<Set<String>, Set<ConnectorHead>> connectorsToMerge = new HashMap<>();
```

Comment:
better to document what this map holds.. specifically what does the key represent?

Reply:
agreed. for the purposes of this review, the key is the set of input nodes to the connector branch head.